test/e2e/logforwarding/fluent/fluent_test.go

```diff
@@ -0,0 +1,154 @@
+package fluent_test
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	loggingv1 "github.com/openshift/cluster-logging-operator/pkg/apis/logging/v1"
+	"github.com/openshift/cluster-logging-operator/test"
+	"github.com/openshift/cluster-logging-operator/test/client"
+	"github.com/openshift/cluster-logging-operator/test/helpers/cmd"
+	"github.com/openshift/cluster-logging-operator/test/helpers/fluentd"
+	. "github.com/openshift/cluster-logging-operator/test/matchers"
+	"github.com/openshift/cluster-logging-operator/test/runtime"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const message = "My life is my message"
+
+var (
+	// Create/delete clusterlogging instance around entire suite.
+	cl = runtime.NewClusterLogging()
+	_  = BeforeSuite(func() { ExpectOK(client.Get().Recreate(cl)) })
+	_  = AfterSuite(func() { _ = client.Get().Remove(cl) })
+)
+
+// Extend client.Test for tests using logGenerator and receiver.
+type Test struct {
+	*client.Test
+	receiver     *fluentd.Receiver
+	logGenerator *corev1.Pod
+	group        test.FailGroup // Run cluster operations in parallel
+}
+
+func (t *Test) Reader(name string) *cmd.Reader {
+	return t.receiver.Sources[name].TailReader()
+}
+
+func NewTest(message string) *Test {
+	ct := client.NewTest()
+	t := &Test{
+		Test:         ct,
+		receiver:     fluentd.NewReceiver(ct.NS.Name, "receiver"),
+		logGenerator: runtime.NewLogGenerator(ct.NS.Name, "log-generator", 10000, 0, message),
+	}
+	t.group.Go(func() { ExpectOK(t.Create(t.logGenerator)) })
+	return t
+}
+
+func (t *Test) Close() {
+	defer t.Test.Close()
+	t.group.Wait()
+}
+
+var _ = Describe("[ClusterLogForwarder]", func() {
+	var t *Test
+	BeforeEach(func() { t = NewTest(message) })
+	AfterEach(func() { t.Close() })
+
+	Context("with app/infra/audit receiver", func() {
+		BeforeEach(func() {
+			t.receiver.AddSource("application", "forward", 24224)
+			t.receiver.AddSource("infrastructure", "forward", 24225)
+			t.receiver.AddSource("audit", "forward", 24226)
+			t.group.Go(func() { ExpectOK(t.receiver.Create(t.Client)) })
```

Comment:
As a team, we had a conversation of terse vs explicit and frankly I find this syntax non-intuitive. Is there something we can do here by name that can make it more obvious what it does?  I don't understand why its structured this way at all

Reply:
How would you like it structured? Here are some other options
```
// Use struct field names to be more explicit
receiver.AddSource("name", Source{Port: 24224, Type: "forward"})
// Provide all sources as map on creation, no Add function
receiver := NewReceiver(map[string]*Source{ "name": Source{Port: 24224, Type: "forward"}, .... }
```
I used "AddSource/Source" because that's what fluentd calls it, personally I would tend to use "Listen/Listener" to follow the go net library. I guess it depends if we want to follow fluent or Go conventions here.

Reply:
Updated using explicit struct fields, I like this better than the original. Whaddya think @jcantrill ?

Reply:
The source was less of a concern to me then `t.group.Go(func() { ExpectOK(t.receiver.Create(t.Client)) })` which doesn't read well to be or make it obvious whats going on.   I'm coming from the perspective of what is it I need to do when I add a test and why am I doing it

Reply:
This line is starting a goroutine to create a recever using t.Client.
go func() {...}() is how you start a goroutine, so 
g.Go(func() {...}) starts a gouroutine managed by the group g. This is syntactically pretty close, and the same pattern as the errgroup package which is a proposed Go extension https://godoc.org/golang.org/x/sync/errgroup.
ExpectOK() is shorthand for Expect(x).To(Succeed()) which *really* annoys me for unnecessry typing in Ginkgo. 
Which part is problematic?